Implemented unsized `Script`

Kixunil opened the pull request:
This renames `Script` to `ScriptBuf` and adds unsized `Script` modeled
after `PathBuf`/`Path`. The change cleans up the API a bit, especially
all functions that previously accepted `&Script` now accept truly
borrowed version. Some functions that perviously accepted `&[u8]` can
now accept `&Script` because constructing it is no loger costly.

This is a more idiomatic alternative to #884 with two unavoidable lines of `unsafe` copied from `std`.

Closes #522 
Closes #949 

(For 949, we allow users to use whichever they like but still use `ScriptBuf` in `Transaction`.)

tcharding commented:
There are a bunch of things I don't fully understand about this code but FWIW concept ACK from me, its a much cleaner that #884.

tcharding commented on src/blockdata/script.rs:
```diff
@@ -51,9 +53,79 @@ use crate::schnorr::{TapTweak, TweakedPublicKey, UntweakedPublicKey};
-#[derive(Clone, Default, PartialOrd, Ord, PartialEq, Eq, Hash)]
-pub struct Script(Box<[u8]>);
+#[derive(PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Script([u8]);
```
As you work on this perhaps include this doc string modified from stdlib

```
+/// This is an unsized type, meaning that it must always be used behind a pointer like & or Box. For
+/// an owned version of this type, see [`ScriptBuf`].
```

Kixunil replied:
Yeah, doc will be improved when I get more concept ACKs since there are two lines of `unsafe`.

tcharding replied:
no worries, just thought I'd throw it in since I was reading up trying to figure out what you were doing :)

Kixunil commented:
Feel free to ask if you need help with understanding!

Kixunil commented:
Sadly, I'm [not 100% sure the casts are sound](https://internals.rust-lang.org/t/is-casting-dst-to-newtype-dst-actually-sound/17116). I believe the conclusion will be "it's OK" but until then it's formally UB.

elichai commented on src/blockdata/script.rs:
```diff
@@ -51,9 +53,79 @@ use crate::schnorr::{TapTweak, TweakedPublicKey, UntweakedPublicKey};
-#[derive(Clone, Default, PartialOrd, Ord, PartialEq, Eq, Hash)]
-pub struct Script(Box<[u8]>);
+#[derive(PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct Script([u8]);
```
Add `#[repr(transparent)]` to allow the casts?

Kixunil replied:
I don't believe it helps. `#[repr(transparent)]` only affects the relationship between newtypes and their encapsulated types, not references to them. However this looks promising: https://github.com/rust-lang/reference/pull/1206

Kixunil replied:
Thanks you! Looks like [`std` has the same issue](https://github.com/rust-lang/rust/blob/792bc5a0102d0973d42183a2b267850bb905236f/library/std/src/path.rs#L1162-L1167) however being `std` it can afford to do what it does and us being external library should do the opposite: use `#[repr(transparent)]` and document it's an implementation detail, right?

tcharding commented on src/blockdata/script.rs:
```diff
@@ -643,107 +678,339 @@ impl Script {
+    }
+}
+
+impl Deref for ScriptBuf {
```
My mind was just blown, this is acting like compile time polymorphism, right? This means that any method on `Script` can be called on a `ScriptBuf` because the compiler automatically dereferences the `ScriptBuf`, that is hot AF.

Kixunil replied:
Not really polymorphism, just `Deref` coercions. They work very well in case of `String` or `Vec` in my experience, so I went with it.

apoelstra replied:
I wouldn't use the word "just" to refer to `Deref` coercions :). I agree that this is one of the weirdest most magical parts of the language.

tcharding commented on src/blockdata/script.rs:
```diff
@@ -643,107 +678,339 @@ impl Script {
-    pub fn asm(&self) -> String {
-        Script::bytes_to_asm(self.as_ref())
+    /// Adds a single opcode to the script.
+    pub fn push_opcode(&mut self, data: opcodes::All) {
+        self.0.push(data.to_u8());
+    }
+
+    /// Adds instructions to push some arbitrary data onto the stack.
+    pub fn push_slice(&mut self, data: &[u8]) {
```
Its not clear to me why these methods are on `ScriptBuf` when we have the `Builder`, can you explain the use case please?

Kixunil replied:
Yes, wanted to document better, but forgot. `Builder` has more ergonomic API for build-and-hand-over code but it requires `self`, where `ScriptBuf` takes `&mut self` instead.

tcharding replied:
Cool, I did not notice that detail. Thanks.

tcharding reviewed (commented):
There is a lot going on in this PR, I'm not sure my knowledge is deep enough to ack but I had a good long look and to the best of my understanding this looks pretty good.

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -51,24 +76,111 @@ use crate::schnorr::{TapTweak, TweakedPublicKey, UntweakedPublicKey};
+    }
+}
+
+impl Script {
```
Wanted to make `unsafe` separated and later I realized it's not needed here and forgot to merge it. :)

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -600,29 +610,54 @@ impl Script {
+
+    /// Iterates over the script instructions and their indices while enforcing minimal pushes.
+    ///
+    /// This is similar to [`instruction_indices`](Self::instruction_indices) but an error is
```
Ah, yes, I solved it in other case saying it's a method.

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -643,107 +678,339 @@ impl Script {
+    }
+
+    /// Generates P2WPKH-type of scriptPubkey.
+    #[deprecated(since = "0.28.0", note = "use ScriptBuf::new_v0_p2wpkh method instead")]
```
Remove them completely rather, I mindlessly moved things around.

Kixunil commented:
Review **not** addressed, just rebased for now. Couldn't find enough time. I'm half-sleeping rn.

tcharding commented:
> I'm half-sleeping rn.

Hell of a team today then. I woke up in the middle of the night for some reason and started coding, pretty sure I will have done some brain fried stuff over the last few hours :)

Kixunil commented on src/lib.rs:
```diff
@@ -33,7 +33,7 @@
 #![cfg_attr(docsrs, feature(doc_cfg))]
 // Coding conventions
-#![forbid(unsafe_code)]
+#![deny(unsafe_code)]
```
Hmm, I just realized that by that logic we shouldn't have depended on `secp256k1` which has "buried `unsafe` code with implicit `allow`". I was thinking of moving `Script{,Buf,Builder}` to a separate crate to contain `unsafe` but this refutes the idea. (Could there be another good reason to move it though?)

Kixunil commented:
Pushed more changes but review not addressed yet.

Kixunil commented:
Last push addressed review comments and added a few more conversion APIs. Unfortunately, some required (also easy) `unsafe`. It's still not as good as `String`/`PathBuf`/... though, that will require a bunch more methods.

tcharding commented on src/address.rs:
```diff
@@ -608,7 +610,7 @@ impl<'a> fmt::Display for AddressEncoding<'a> {
 /// ### Relevant BIPs
 ///
 /// * [BIP13 - Address Format for pay-to-script-hash](https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki)
-/// * [BIP16 - Pay to Script Hash](https://github.com/bitcoin/bips/blob/master/bip-0016.mediawiki)
+/// * [BIP16 - Pay to ScriptBuf Hash](https://github.com/bitcoin/bips/blob/master/bip-0016.mediawiki)
```
Should be `Script` still.

Kixunil replied:
LOL, I did search&replace and I didn't realize the compiler won't complain if the doc is wrong.

tcharding commented on src/address.rs:
```diff
@@ -63,7 +63,7 @@ pub enum Error {
         found: bech32::Variant,
     },
-    /// Script version must be 0 to 16 inclusive.
+    /// ScriptBuf version must be 0 to 16 inclusive.
```
Should be `Script` still.

tcharding commented on src/address.rs:
```diff
@@ -77,7 +77,7 @@ pub enum Error {
     /// Address size more than 520 bytes is not allowed.
     ExcessiveScriptSize,
-    /// Script is not a p2pkh, p2sh or witness program.
+    /// ScriptBuf is not a p2pkh, p2sh or witness program.
```
Should be `Script` still.

tcharding commented on src/blockdata/script.rs:
```diff
@@ -3,72 +3,552 @@
+/// The type is `#[repr(transparent)]` for internal purposes only!
+/// No consumer crate may rely on the represenation of the struct!
+///
+/// ## References
+///
+///
+/// ### Bitcoin Core References
```
Do we need both these headings? If so, we could get rid of line 122

tcharding commented on src/blockdata/script.rs:
```diff
@@ -3,72 +3,552 @@
+    /// Treat byte slice as `Script`
+    #[inline]
+    pub fn from_bytes(bytes: &[u8]) -> &Script {
+        unsafe {
```
Perhaps document the `unsafe`s.

Kixunil replied:
There's a clippy lint for that, we should activate it.

tcharding replied:
Seems it only lints for unsafe functions, not functions that include unsafe blocks internally. Its on by default.

tcharding commented on src/blockdata/script.rs:
```diff
@@ -324,20 +796,122 @@ fn read_uint_iter(data: &mut core::slice::Iter<'_, u8>, size: usize) -> Result<u
+    }
+
+    /// Returns a mutable reference to unsized script.
+    pub fn as_mut_script(&mut self) -> &mut Script {
```
Perhaps `as_script_mut` is better? I think I read somewhere to put `mut` on the end but I failed to find a reference so not totally sure.

tcharding replied:
Oh and we have `as_bytes` and `as_bytes_mut` so would be uniform with those.

Kixunil replied:
There's [as_mut_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_slice) I modeled it after that - it's the analogous function. I need to change `as_bytes_mut` instead.

tcharding replied:
Cool, TIL. `as_mut_ptr` exists too. Cheers.

tcharding commented on src/blockdata/script.rs:
```diff
@@ -324,20 +796,122 @@ fn read_uint_iter(data: &mut core::slice::Iter<'_, u8>, size: usize) -> Result<u
+
+impl ScriptBuf {
     /// Creates a new empty script.
-    pub fn new() -> Script { Script(vec![].into_boxed_slice()) }
+    pub fn new() -> Self {
+        ScriptBuf(Vec::new())
+    }
+
+    /// Creates a new empty script with pre-allocated capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        ScriptBuf(Vec::with_capacity(capacity))
+    }
+
+    /// Pre-allocates `additional_len` bytes if needed.
+    pub fn reserve(&mut self, additional_len: usize) {
+        self.0.reserve(additional_len);
+    }
+
+    /// Returns a reference to unsized script.
+    pub fn as_script(&self) -> &Script {
+        Script::from_bytes(&self.0)
+    }
+
+    /// Returns a mutable reference to unsized script.
+    pub fn as_mut_script(&mut self) -> &mut Script {
+        Script::from_bytes_mut(&mut self.0)
+    }
 
     /// Generates P2PK-type of scriptPubkey.
-    pub fn new_p2pk(pubkey: &PublicKey) -> Script {
+    pub fn new_p2pk(pubkey: &PublicKey) -> Self {
         Builder::new()
             .push_key(pubkey)
             .push_opcode(opcodes::all::OP_CHECKSIG)
             .into_script()
     }
 
     /// Generates P2PKH-type of scriptPubkey.
-    pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Script {
+    pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Self {
```
With two types so similar in the same file perhaps its clearer to return `Script` instead of `Self`?

Kixunil replied:
I always use `Self` for return types to make refactorings less painful. Although ideally there won't be any such refactorings...

tcharding commented on src/util/psbt/map/input.rs:
```diff
@@ -33,9 +33,9 @@ const PSBT_IN_WITNESS_UTXO: u8 = 0x01;
 /// Type: Sighash Type PSBT_IN_SIGHASH_TYPE = 0x03
 const PSBT_IN_SIGHASH_TYPE: u8 = 0x03;
-/// Type: Redeem Script PSBT_IN_REDEEM_SCRIPT = 0x04
+/// Type: Redeem ScriptBuf PSBT_IN_REDEEM_SCRIPT = 0x04
```
This and below should all be `Script`

stevenroose commented on src/blockdata/script.rs:
```diff
@@ -3,72 +3,565 @@
+impl Script {
+    /// Treat byte slice as `Script`
+    #[inline]
+    pub fn from_bytes(bytes: &[u8]) -> &Script {
```
I never noticed how this was possible! So there is a way to do this with unsafe! Amazing!

stevenroose commented on src/consensus/encode.rs:
```diff
@@ -673,6 +673,13 @@ fn read_bytes_from_finite_reader<D: io::Read>(mut d: D, mut opts: ReadBytesFromF
     Ok(ret)
 }
 
+impl Encodable for [u8] {
```
I would avoid doing this. It makes it quite error-prone, because it's unclear if slices would be length-prefixed or not. If we use this internally, I'd prefer using `consensus_encode_with_size` instead..

stevenroose replied:
Generally, we just use the `emit_slice` method from `WriteExt`.

Kixunil replied:
Good point, though IIRC `Vec<u8>` implements `Encodable` and it shouldn't for the same reason.

tcharding commented:
Whats the status of this one @Kixunil?

Kixunil commented:
Definitely want to continue but since we were doing other things I didn't have much time. Now this will need an insane rebase, unfortunately.

sanket1729 reviewed (commented):
Did a pass-through. Looks good to me, light code review ACK.

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -357,393 +944,325 @@ impl Script {
+        let mut iter = iter.into_iter();
+        // Most of Bitcoin scripts have only a few opcodes, so we can avoid reallocations in many
+        // cases.
+        if iter.size_hint().1.map(|max| max < 6).unwrap_or(false) {
```
Wow, I completely forgot I wrote this. :D I believe I've looked at common scripts and picked the largest which is indeed P2PKH.

Kixunil commented:
I've added milestone as I'm hoping to get it into the next version. I would like to rebase this no later than during the next week.

tcharding commented:
I see no reason we cannot get this in 0.30.0

apoelstra commented:
I agree, I'd really like to get this into the next version.

tcharding commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -1056,31 +1714,47 @@ impl Builder {
+        &self.0
+    }
+
+    /// Returns script bytes 
```
TIL: Our pre-commit hook and clippy CI job allows one character of trailing whitespace - does that mean one character of trailing whitespace is ok, clippy is wrong, or we have the clippy incantation wrong?

`cargo clippy --all-features --all-targets -- -D warnings`

Kixunil replied:
Oh weird, I though I removed all of them.

tcharding commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -3,75 +3,606 @@
+#[cfg_attr(docsrs, doc(cfg(feature = "rust_v_1_53")))]
+delegate_index!((Bound<usize>, Bound<usize>));
 
+impl AsRef<Script> for Script {
```
What is the point of this impl? Is it just for uniformity (along with `as_mut`) so `ScriptBuf` and `Script` both have these implemented?

Kixunil replied:
It allows you to pass `&Script` to functions requiring `AsRef<Script>`. `std` cannot implement this as a blanket impl due to coherence issues (it does impl `AsRef<T> for &T` though) so types need to do it themselves. This is opposite of `Borrow` which has blanket `Borrow<T> for T` but not for references.

[`std` types have this impl too](https://doc.rust-lang.org/stable/std/path/struct.Path.html#impl-AsRef%3CPath%3E-for-Path) so it's also consistent with that.

tcharding commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -453,378 +1057,335 @@ impl Script {
+
+impl Borrow<Script> for ScriptBuf {
+    fn borrow(&self) -> &Script {
+        self
```
Can you explain to me how this works please? I was expecting to see `self.as_script()`. Same question for all the others (especially `AsRef`).

Kixunil replied:
Deref coercion :)

tcharding reviewed (approved):
Excluding a few minor things that I don't quite get, ACK a620afb0d767b380b505ccc8a5a51ba0b2212b3a

Kixunil commented:
I went over the PR and noticed/fixed some more issues. One still remains: `Display` calls into `Debug` which outputs `Script(ASM)`. Firstly, calling `Debug` from `Display` is wrong. But what should `Script` even display? No representation (hex, asm) seems obvious enough.

Because of the confusion I'd prefer removing `Display` and adding `display_hex()` and `display_asm()` methods similar to how we handled amount.

apoelstra commented:
I think `Display` should show ASM and we should use the `LowerHex`/`UpperHex` traits for hex. The reason being that while either hex or asm could be the "natural" choice for `Dispaly`, hex gets its own format specifier so we can just move it there.

I don't think it'll cause confusion in practice; output of `Display` is the *one* thing that programmers visually check :P.

Kixunil commented:
Oh, yes, I'm dumb last couple of days (I'm a bit sick).
Should we put `Script()` around it though? Seems wrong for `Display`.

apoelstra commented:
Agreed it seems wrong -- so I vote no on putting `Script` around it.

apoelstra commented:
> I think `Display` should show ASM and we should use the `LowerHex`/`UpperHex` traits for hex. The reason being that while either hex or asm could be the "natural" choice for `Dispaly`, hex gets its own format specifier so we can just move it there.
> 
> I don't think it'll cause confusion in practice; output of `Display` is the _one_ thing that programmers visually check :P.

@Kixunil I think that if you implement this, and rebase, I can do a final review of this PR and we can finally get it in.

Kixunil commented:
I noticed hex encoding is using the slow implementation. Since I'm low on time and this PR may block others I suggest to let it in and I will change to fast impl in a subsequent PR. Also I kept `Script()` in `Debug` impl. Again, if we want to change it I suggest a separate PR.

Kixunil commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -3,75 +3,639 @@
 
+delegate_index!(Range<usize>, RangeFrom<usize>, RangeTo<usize>, RangeFull, RangeInclusive<usize>, RangeToInclusive<usize>);
+#[cfg(feature = "rust_v_1_53")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rust_v_1_53")))]
```
FYI this doc attribute doesn't work. I prefer to fix it in a separate PR.

apoelstra commented:
Needs rebase again, sorry, I think because of #1330.

apoelstra commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -457,378 +1120,345 @@ impl Script {
+    ///
+    /// Note that existing `OP_*VERIFY` opcodes do not lead to the instruction being ignored
+    /// because `OP_VERIFY` consumes an item from the stack so ignoring them would change the
+    /// semantics.
```
nit: (can be dealt with in a followup PR). I don't understand this sentence.

Kixunil replied:
It's he issue I brought up before :) I will try to think of better wording after I sleep.

apoelstra replied:
Maybe "Note that `VERIFY` itself is not a `*VERIFY` opcode -- repeated instances of `OP_VERIFY` will each pop a distinct stack element, so they cannot be combined."

Anyway we can debate this on a followup PR, let's not hold up a 2000-line diff on this :)

apoelstra commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -457,378 +1120,345 @@ impl Script {
+    /// Adds an `OP_VERIFY` to the script or changes the most-recently-added opcode to `VERIFY`
+    /// alternative.
+    ///
+    /// See the public fn [`Self::scan_and_push_verify`] to learn more.
```
nit: we should mention here that we don't check whether `last_opcode` is actually the last opcode

Kixunil replied:
This is a private method so it shouldn't matter that much but yeah, something for another PR.

apoelstra replied:
Oh, perfect :) for a private method it's fine, it's obvious from the code that it doesn't check.

apoelstra commented:
Wow, what a magnum opus! This is very thorough. I appreciate all the long-form documentation as well as the many utility methods that make `ScriptBuf` ergonomic. ACK.

apoelstra reviewed (approved):
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03

tcharding reviewed (approved):
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03

tcharding commented on bitcoin/src/lib.rs:
```diff
@@ -59,7 +59,6 @@ compile_error!(
 #[cfg(bench)]
 extern crate test;
 
-#[cfg(feature = "no-std")]
```
I don't fully understand the cost/benefits of always using the `alloc` crate vs using std for std builds and alloc only for no-std builds. Not important for this PR but its something I've noticed I didn't understand this week, if you have time ever can you explain your thoughts please?

tcharding replied:
Oh, needs more context, the lines below this are 
```rust
#[macro_use]
extern crate alloc;
```

Kixunil replied:
> I don't fully understand the cost/benefits of always using the alloc crate vs using std for std builds and alloc only for no-std builds. 

Just simpler code, less compilation conditions to worry about.

sanket1729 commented:
Please wait before merging this. On top of my list for reviews today. Should be done in a couple of hours.

sanket1729 reviewed (approved):
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03. Let's get this in

sanket1729 commented:
I think 3 ACKs should this good to go. Merging this to prevent other accidental merges from breaking this big PR

sanket1729 commented:
and BOOM!

Kixunil commented:
Thanks a lot folks for quick review, I didn't expect it!
